File: icebeach-wakeclub/apps/dashboard/src/lib/gender.ts
import { WetsuitGender } from "../types";

const FEMALE_NAMES = new Set([
  "александра", "алена", "алина", "алла", "анастасия", "анна", "валентина", "валерия",
  "варвара", "вера", "вероника", "виктория", "галина", "дарья", "диана", "евгения",
  "екатерина", "елена", "елизавета", "жанна", "инна", "ирина", "карина", "кристина",
  "ксения", "лариса", "любовь", "людмила", "марина", "мария", "милана", "надежда",
  "наталья", "оксана", "ольга", "полина", "светлана", "снежана", "софия", "софья",
  "тамара", "татьяна", "ульяна", "юлия", "яна",
]);

const MALE_NAMES = new Set([
  "александр", "алексей", "андрей", "антон", "артём", "артем", "борис", "вадим",
  "валентин", "валерий", "виктор", "виталий", "владимир", "владислав", "геннадий",
  "георгий", "григорий", "дмитрий", "евгений", "иван", "игорь", "илья", "кирилл",
  "константин", "лев", "леонид", "максим", "михаил", "никита", "николай", "олег",
  "павел", "пётр", "петр", "рома", "роман", "сергей", "станислав", "степан",
  "тимофей", "фёдор", "федор", "юрий", "ярослав",
]);

const MALE_A_EXCEPTIONS = new Set(["никита", "илья", "савва", "кузьма", "фома", "данила", "саша"]);

function normalizeNamePart(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/ё/g, "е")
    .replace(/[^a-zа-я-]/gi, "");
}

export function inferGenderFromFullName(fullName: string): WetsuitGender | null {
  const parts = fullName.split(/\s+/).map(normalizeNamePart).filter(Boolean);
  if (parts.length === 0) {
    return null;
  }

  // Имя может стоять первым или вторым: «Ирина Смирнова» и «Смирнова Ирина».
  for (const part of parts) {
    if (FEMALE_NAMES.has(part)) {
      return "female";
    }
  }
  for (const part of parts) {
    if (MALE_NAMES.has(part)) {
      return "male";
    }
  }

  for (const part of parts) {
    if (/(овна|евна|ична)$/.test(part)) {
      return "female";
    }
    if (/(ович|евич)$/.test(part)) {
      return "male";
    }
  }

  if (parts.length > 1) {
    for (const part of parts) {
      if (/(ова|ева|ина|ына|ая|ская)$/.test(part)) {
        return "female";
      }
      if (/(ов|ев|ин|ын|ский|цкий)$/.test(part)) {
        return "male";
      }
    }
  }

  const first = parts[0];
  if (MALE_A_EXCEPTIONS.has(first)) {
    return "male";
  }
  if (/[ая]$/.test(first)) {
    return "female";
  }
  if (/[бвгджзклмнпрстфхцчшщй]$/.test(first)) {
    return "male";
  }
  return null;
}

export function genderLabel(gender?: WetsuitGender | null): string {
  if (gender === "male") return "Муж";
  if (gender === "female") return "Жен";
  return "—";
}

export function displayGenderForClient(fullName: string, stored?: WetsuitGender | null): string {
  return genderLabel(stored || inferGenderFromFullName(fullName));
}

File: icebeach-wakeclub/apps/api/app/services/gender.py
from __future__ import annotations

FEMALE_NAMES = {
    "александра",
    "алена",
    "алина",
    "алла",
    "анастасия",
    "анна",
    "валентина",
    "валерия",
    "варвара",
    "вера",
    "вероника",
    "виктория",
    "галина",
    "дарья",
    "диана",
    "евгения",
    "екатерина",
    "елена",
    "елизавета",
    "жанна",
    "инна",
    "ирина",
    "карина",
    "кристина",
    "ксения",
    "лариса",
    "любовь",
    "людмила",
    "марина",
    "мария",
    "милана",
    "надежда",
    "наталья",
    "оксана",
    "ольга",
    "полина",
    "светлана",
    "снежана",
    "софия",
    "софья",
    "тамара",
    "татьяна",
    "ульяна",
    "юлия",
    "яна",
}

MALE_NAMES = {
    "александр",
    "алексей",
    "андрей",
    "антон",
    "артем",
    "борис",
    "вадим",
    "валентин",
    "валерий",
    "виктор",
    "виталий",
    "владимир",
    "владислав",
    "геннадий",
    "георгий",
    "григорий",
    "дмитрий",
    "евгений",
    "иван",
    "игорь",
    "илья",
    "кирилл",
    "константин",
    "лев",
    "леонид",
    "максим",
    "михаил",
    "никита",
    "николай",
    "олег",
    "павел",
    "петр",
    "рома",
    "роман",
    "сергей",
    "станислав",
    "степан",
    "тимофей",
    "федор",
    "юрий",
    "ярослав",
}

MALE_A_EXCEPTIONS = {"никита", "илья", "савва", "кузьма", "фома", "данила", "саша"}


def _normalize_name_part(value: str) -> str:
    cleaned = value.strip().lower().replace("ё", "е")
    return "".join(char for char in cleaned if char.isalpha() or char == "-")


def infer_gender_from_full_name(full_name: str) -> str | None:
    """Guess male/female from a Russian full name when Sheets has no wetsuit gender."""
    parts = [_normalize_name_part(part) for part in full_name.split() if _normalize_name_part(part)]
    if not parts:
        return None

    for part in parts:
        if part in FEMALE_NAMES:
            return "female"
    for part in parts:
        if part in MALE_NAMES:
            return "male"

    for part in parts:
        if part.endswith(("овна", "евна", "ична")):
            return "female"
        if part.endswith(("ович", "евич")):
            return "male"

    if len(parts) > 1:
        for part in parts:
            if part.endswith(("ова", "ева", "ина", "ына", "ая", "ская")):
                return "female"
            if part.endswith(("ов", "ев", "ин", "ын", "ский", "цкий")):
                return "male"

    first = parts[0]
    if first in MALE_A_EXCEPTIONS:
        return "male"
    if first.endswith(("а", "я")):
        return "female"
    if first and first[-1] in "бвгджзклмнпрстфхцчшщй":
        return "male"
    return None
